File: CodingInterviewsPackSolutionsCPP/4Sum.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        int n = nums.size();
        sort(nums.begin(),nums.end());
        vector<vector<int>>result;
        set<vector<int>>filter;
        
        for(int i = 0; i <= n-4; i++){
            for(int j = i+1; j<=n-3; j++){
                int left = j+1;
                int right = n-1;
                
                while(left<right){
                    long long sum = nums[i] + nums[j];
                    sum += nums[left];
                    sum += nums[right] - target;
                    if(!sum){
                        vector<int>v({nums[i],nums[j],nums[left],nums[right]});
                        sort(v.begin(),v.end());
                        if(filter.insert(v).second)
                            result.push_back(v);
                        left++;
                        right--;
                    }else if(sum>0)
                        right--;
                    else
                        left++;
                }
            }
        }
        return result;
    }
};

File: CodingInterviewsPythonSolutions/4Sum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        n = len(nums)
        nums.sort()
        result = []
        filter_set = set()
        
        for i in range(n - 3):
            for j in range(i + 1, n - 2):
                left = j + 1
                right = n - 1
                
                while left < right:
                    sum_val = nums[i] + nums[j] + nums[left] + nums[right] - target
                    
                    if sum_val == 0:
                        quad = [nums[i], nums[j], nums[left], nums[right]]
                        quad.sort()
                        
                        if tuple(quad) not in filter_set:
                            result.append(quad)
                            filter_set.add(tuple(quad))
                        
                        left += 1
                        right -= 1
                    elif sum_val > 0:
                        right -= 1
                    else:
                        left += 1
        
        return result
